# 0. Tools and Python (3.7.4) packages   
**bedtools v2.31.0**  
**pandas v1.3.5**   
**NumPy v1.21.5**  
**Bio v1.5.9** 

# Resources   
1. [UCSC Genome Browser](http://genome.ucsc.edu/index.html)   
2. [NCBI](https://www.ncbi.nlm.nih.gov/)




In [1]:
import pandas as pd
import numpy as np
from Bio import Entrez
Entrez.email = "an.eudokimova@yandex.ru"

****************

# 1. Task1. Annotation of input BED-file   
1.1 Download GTF files for the main gene transcript sets from [[1]]. The BED-file header states that reference=hg19:     
```bash
wget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/genes/hg19.ncbiRefSeq.gtf.gz
```   
```bash 
gunzip hg19.ncbiRefSeq.gtf.gz
```   
   
Based on the wording of the task, we are only interested in exons, so we extract only exons from GTF-file:   
```bash 
grep -w 'exon' hg19.ncbiRefSeq.gtf > hg19.ncbiRefSeq.exon.gtf
```
   
1.2 To identify gene regions in input BED-file we will use bedtools intersect, this tool allows one to screen for overlaps between two sets of genomic features:   
``` bash 
bedtools intersect -a IAD143293_241_Designed.bed -b hg19.ncbiRefSeq.exon.gtf -loj -wa -wb > intersect.txt
```   
Options `-wa` and `-wb` allows to write the original entries of both A and B input files in output   
Option `-loj` report a NULL feature for B, if no overlaps are found   

Creat directory for final results:   
```bash 
mkdir RESULTS
```

[1]: https://hgdownload.soe.ucsc.edu/


1.3 Analisys of output file   

In [2]:
intersect = pd.read_csv('intersect.txt', sep='\t', header=None)
intersect.head(5)

0          1          2               3  4   \
0  chr1   18722630   18722936  AMPL7162509462  .   
1  chr1  113456395  113456721  AMPL7166063003  .   
2  chr1  113456395  113456721  AMPL7166063003  .   
3  chr1  113459897  113460203  AMPL7166063010  .   
4  chr1  113459897  113460203  AMPL7166063010  .   

                                       5     6                      7     8   \
0  Pool=1;SUBMITTED_REGION=AMPL7162509462     .                      .     .   
1  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   
2  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   
3  Pool=1;SUBMITTED_REGION=AMPL7166063010  chr1  ncbiRefSeq.2021-05-17  exon   
4  Pool=1;SUBMITTED_REGION=AMPL7166063010  chr1  ncbiRefSeq.2021-05-17  exon   

          9          10 11 12 13  \
0         -1         -1  .  .  .   
1  113454469  113456787  .  -  .   
2  113454469  113456787  .  -  .   
3  113459800  113460666  .  -  .   
4  113459800  113460666  .  -  .   

                                                  14  
0                                                  .  
1  gene_id "SLC16A1"; transcript_id "NM_001166496...  
2  gene_id "SLC16A1"; transcript_id "NM_003051.4"...  
3  gene_id "SLC16A1"; transcript_id "NM_001166496...  
4  gene_id "SLC16A1"; transcript_id "NM_003051.4"...

We can see, that not all amplicons are identified (-1 in columns 9 and 10).    
Extraction gene name and exon number from column 14:  

In [3]:
tmp = intersect[14].str.split(' ', expand=True).iloc[:, [9, 5]]. \
    reset_index().drop('index', axis=1).replace(np.nan, '.').replace([';', '"'], '', regex=True)
tmp.columns = [15, 16]
out = pd.concat([intersect, tmp], axis=1)
out.head(6)

0          1          2               3  4   \
0  chr1   18722630   18722936  AMPL7162509462  .   
1  chr1  113456395  113456721  AMPL7166063003  .   
2  chr1  113456395  113456721  AMPL7166063003  .   
3  chr1  113459897  113460203  AMPL7166063010  .   
4  chr1  113459897  113460203  AMPL7166063010  .   
5  chr1  113460270  113460593  AMPL7163702396  .   

                                       5     6                      7     8   \
0  Pool=1;SUBMITTED_REGION=AMPL7162509462     .                      .     .   
1  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   
2  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   
3  Pool=1;SUBMITTED_REGION=AMPL7166063010  chr1  ncbiRefSeq.2021-05-17  exon   
4  Pool=1;SUBMITTED_REGION=AMPL7166063010  chr1  ncbiRefSeq.2021-05-17  exon   
5  Pool=2;SUBMITTED_REGION=AMPL7163702396  chr1  ncbiRefSeq.2021-05-17  exon   

          9          10 11 12 13  \
0         -1         -1  .  .  .   
1  113454469  113456787  .  -  .   
2  113454469  113456787  .  -  .   
3  113459800  113460666  .  -  .   
4  113459800  113460666  .  -  .   
5  113459800  113460666  .  -  .   

                                                  14       15 16  
0                                                  .        .  .  
1  gene_id "SLC16A1"; transcript_id "NM_001166496...  SLC16A1  5  
2  gene_id "SLC16A1"; transcript_id "NM_003051.4"...  SLC16A1  5  
3  gene_id "SLC16A1"; transcript_id "NM_001166496...  SLC16A1  4  
4  gene_id "SLC16A1"; transcript_id "NM_003051.4"...  SLC16A1  4  
5  gene_id "SLC16A1"; transcript_id "NM_001166496...  SLC16A1  4

We can see rows with the same amplicons, gene names and exon coordinates and number, but with different transcript_id in column 14. As far as we are not interested in transcript_id, we can remove this duplicates:

In [4]:
out = out.drop_duplicates([3, 15, 16], keep='first').reset_index().drop('index', axis=1)

...and finaly extract target columns, rename them and write output file:

In [5]:
out = out.iloc[:, [0,1, 2, 3, 4, 5, 15, 16]]
out.columns = ['#CHR', 'START_POS', 'END_POS', 'AMPL_ID', 'STRAND', 'AMPL_DESCRIPTION', 'GENE_NAME', 'EXON_NUMBER']
out.to_csv('RESULTS/task1_output.bed', sep='\t', index=None, header=True)
out.head(6)

#CHR  START_POS    END_POS         AMPL_ID STRAND  \
0  chr1   18722630   18722936  AMPL7162509462      .   
1  chr1  113456395  113456721  AMPL7166063003      .   
2  chr1  113459897  113460203  AMPL7166063010      .   
3  chr1  113460270  113460593  AMPL7163702396      .   
4  chr1  113471727  113472054  AMPL7166063018      .   
5  chr1  113498778  113499101  AMPL7156752826      .   

                         AMPL_DESCRIPTION    GENE_NAME EXON_NUMBER  
0  Pool=1;SUBMITTED_REGION=AMPL7162509462            .           .  
1  Pool=2;SUBMITTED_REGION=AMPL7166063003      SLC16A1           5  
2  Pool=1;SUBMITTED_REGION=AMPL7166063010      SLC16A1           4  
3  Pool=2;SUBMITTED_REGION=AMPL7163702396      SLC16A1           4  
4  Pool=1;SUBMITTED_REGION=AMPL7166063018      SLC16A1           2  
5  Pool=2;SUBMITTED_REGION=AMPL7156752826  SLC16A1-AS1           1

Add header of input BED-file in output file:   


In [6]:
with open('IAD143293_241_Designed.bed', 'r') as input:
    header = input.readline()  
with open('RESULTS/task1_output.bed', 'r+') as output:
    content = output.read()
    output.seek(0, 0)  
    output.write(header)  
    output.write(content)  

1.4 Get information about genes

In [7]:
# function for searching gene ID, description and summary
def get_id(gene):
    try:
        handle = Entrez.esearch(db="gene",term=f"Homo[Orgn] AND {gene}[Gene]")
        result = Entrez.read(handle)
        to_return = result['IdList'][0]
    except:
        to_return = np.nan
    return to_return

def get_summary(ID):
    try:
        handle = Entrez.epost("gene", id=ID)
        result = Entrez.read(handle)
        webEnv = result["WebEnv"]
        queryKey = result["QueryKey"]
        data = Entrez.esummary(db="gene", webenv=webEnv, query_key=queryKey)
        result = Entrez.read(data)
        to_return = [result['DocumentSummarySet']['DocumentSummary'][0]['Description'],
            result['DocumentSummarySet']['DocumentSummary'][0]['Summary']]
    except:
        to_return = [np.nan, np.nan]
    return to_return

In [8]:
# create DF with gene names
genes = pd.DataFrame({'GENE_NAME':out['GENE_NAME'].unique()})
genes = genes[genes['GENE_NAME'] != '.'].reset_index().drop('index', axis=1)

# get ID, description and summary of each gene using Entrez
genes['ID'] = genes['GENE_NAME'].apply(get_id)
tmp = pd.DataFrame(list(map(get_summary, genes['ID'])))
tmp.columns = ['DESCRIPTION', 'SUMMARY']
genes = pd.concat([genes, tmp], axis=1)

genes.to_csv('RESULTS/genes_summary.csv', sep='\t', index=None)


We have 31 genes (see table below). Some functions of protein encoded by these genes:   
- transpoters (*SLC16A1; ABCC8*)   
- enzymes (*G6PC2; HADH*)   
- transcription regulatory factors or part of them (*KLF11 - cell growth inhibition and apoptosis; NEUROD1 - regulation expression of the insulin gene, ZFP57; RFX6; PAX4; GLIS3; NEUROG3 - involved in neurogenesis; HNF1A - regulation of expression of several liver-specific genes; HNF4A; PDX1 - activator of insulin; somatostatin; glucokinase etc; HNF1B - regulation of nephron and embryonic pancreas development; FOXP3*)   
- kinases (*GCK; BLK - cell proliferation and differentiation; GLUD1 - regulating amino acid-induced insulin secretion; INSR - insulin receptor*)   
- potassium channel (*KCNJ11*)    
    
Based on information about these genes, I can assume, that this targeted panel under consideration can be used to diagnose and assess susceptibility to various types of diabetes, as well as other metabolic disorders associated with the functioning of the pancreas and liver.    
Some of these genes can be used for canser diagnostic (for example *BLK, HNF1B, PAX4*), but I don't think that it is a purpose of this pannel

In [9]:
pd.set_option('display.max_colwidth', None)
genes

GENE_NAME         ID  \
0       SLC16A1       6566   
1   SLC16A1-AS1  100506392   
2         KLF11       8462   
3         G6PC2      57818   
4       NEUROD1       4760   
5        MFF-DT     654841   
6          HADH       3030   
7         ZFP57     346171   
8          RFX6     222546   
9           GCK       2645   
10         PAX4       5078   
11          BLK        640   
12        GLIS3     169792   
13    GLIS3-AS1      84850   
14          CEL       1056   
15      NEUROG3      50674   
16        GLUD1       2746   
17          INS       3630   
18     INS-IGF2     723961   
19       KCNJ11       3767   
20        ABCC8       6833   
21        HNF1A       6927   
22     C12orf43      64897   
23         PDX1       3651   
24        HNF1B       6928   
25         INSR       3643   
26       NKX2-2       4821   
27        HNF4A       3172   
28       DBNDD2      55861   
29        FOXP3      50943   
30          UTY       7404   

                                                                      DESCRIPTION  \
0                                               solute carrier family 16 member 1   
1                                                         SLC16A1 antisense RNA 1   
2                                                     KLF transcription factor 11   
3                                       glucose-6-phosphatase catalytic subunit 2   
4                                                      neuronal differentiation 1   
5                                                        MFF divergent transcript   
6   hydroxyacyl-CoA dehydrogenase trifunctional multienzyme complex subunit alpha   
7                                                       ZFP57 zinc finger protein   
8                                                            regulatory factor X6   
9                                                                     glucokinase   
10                                                                   paired box 4   
11                                 BLK proto-oncogene, Src family tyrosine kinase   
12                                                      GLIS family zinc finger 3   
13                                                          GLIS3 antisense RNA 1   
14                                                          carboxyl ester lipase   
15                                                                   neurogenin 3   
16                                                      glutamate dehydrogenase 1   
17                                                                        insulin   
18                                                           INS-IGF2 readthrough   
19                    potassium inwardly rectifying channel subfamily J member 11   
20                                      ATP binding cassette subfamily C member 8   
21                                                                HNF1 homeobox A   
22                                            chromosome 12 open reading frame 43   
23                                             pancreatic and duodenal homeobox 1   
24                                                                HNF1 homeobox B   
25                                                               insulin receptor   
26                                                                 NK2 homeobox 2   
27                                              hepatocyte nuclear factor 4 alpha   
28                                                  dysbindin domain containing 2   
29                                                                forkhead box P3   
30         ubiquitously transcribed tetratricopeptide repeat containing, Y-linked   

                                                                                                                                                                                                                                                                                                                                    

*********************

# 2. Task2. Searching of off-target region in the pannel    
2.1. First of all, let's **repeat BED-file annotation** whithout "exon" filter:   
```bash
bedtools intersect -a IAD143293_241_Designed.bed -b hg19.ncbiRefSeq.gtf -loj -wa -wb > intersect_full.txt
```   
Extract gene names and drop non-overlapping amplicons:

In [10]:
intersect_full = pd.read_csv('intersect_full.txt', sep='\t', header=None)
tmp = intersect_full[14].str.split(' ', expand=True).iloc[:, 9]. \
    reset_index().drop('index', axis=1).replace(np.nan, '.').replace([';', '"'], '', regex=True)
# drop column 14 for shorter view below
df = intersect_full.drop([14], axis=1)
tmp.columns = [14]
df = pd.concat([df, tmp], axis=1)
df = df.loc[df[14] != '.'].reset_index().drop('index', axis=1)

df.head(3)

0          1          2               3  4  \
0  chr1  113456395  113456721  AMPL7166063003  .   
1  chr1  113456395  113456721  AMPL7166063003  .   
2  chr1  113456395  113456721  AMPL7166063003  .   

                                        5     6                      7     8  \
0  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   
1  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  3UTR   
2  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17   CDS   

           9         10 11 12 13       14  
0  113454469  113456787  .  -  .  SLC16A1  
1  113454469  113456512  .  -  .  SLC16A1  
2  113456516  113456787  .  -  2  SLC16A1

In [11]:
len(df[14].unique())

31

2.2 We still have 31 genes. Let's **find out, which amplicons mapped on two or more genes**: 

In [12]:
def find_interesting_genes(df):
    ampls = df[3].unique()
    d = {}
    for ampl in ampls:
        set_genes = list(df.loc[df[3] == ampl][14].unique())
        if len(set_genes) > 1:
            d[ampl] = set_genes
    return d
find_interesting_genes(df)

{'AMPL7156752826': ['SLC16A1-AS1', 'SLC16A1'],
 'AMPL7154848931': ['GLIS3-AS1', 'GLIS3'],
 'AMPL7154784953': ['INS', 'INS-IGF2'],
 'AMPL7154784954': ['INS', 'INS-IGF2'],
 'AMPL7154784955': ['INS', 'INS-IGF2'],
 'AMPL7154897326': ['INS', 'INS-IGF2'],
 'AMPL7154511317': ['C12orf43', 'HNF1A'],
 'AMPL7154784910': ['C12orf43', 'HNF1A'],
 'AMPL7154784911': ['C12orf43', 'HNF1A'],
 'AMPL7154784912': ['C12orf43', 'HNF1A'],
 'AMPL7154783021': ['C12orf43', 'HNF1A'],
 'AMPL7159709172': ['C12orf43', 'HNF1A'],
 'AMPL7159709205': ['C12orf43', 'HNF1A']}

For next steps we take only these genes and amplicons:

In [13]:
tmp_d = find_interesting_genes(df)
tmp_df = pd.DataFrame()
for ampl in tmp_d.keys():
    tmp_df = pd.concat([tmp_df, df.loc[(df[3] == ampl) & df[14].isin(tmp_d[ampl])]], axis=0)

df = tmp_df.reset_index().drop('index', axis=1)

2.3 We already find information about them, but we can **check again, if any of genes are pseudogenes**. For this purpose we use [List] of pseudogenes in hg19. As we expected, there are no these genes in our target pannel.      
But we already can see some "strange" genes in the table above: chromosome 12 open reading frame 43 *C12orf43*, *GLIS3* antisense RNA 1, *SLC16A1* antisense RNA 1, *MFF* divergent transcript. They could be potential off-tartget reogions.   
     
2.4 **Search for homologous regions - 1st way**   
I have three ideas how to search how to search homologous or identical region, but only one I could implement. This is "manualy" method, and it is only suitable for searching for 100% homology.     
     
There are four possible position amplicon on target region (see the picture below): inside (<span style="color:blue">**blue**</span>), partly on left side (<span style="color:green">**green**</span>), partly on right side (<span style="color:yellow">**yellow**</span> ), target inside amplicon (<span style="color:purple">**purple**</span>).   
    
<img src="picture1.jpg" width="100%" height="50%"  />      
   
We can verify this by calculating the lengths of the overlaps, they will be different and often different from the length of the amplicon. If one amplicon mapped on two or more different genome region, we can we can consider 100% homologous overlap sequencies in three cases:   
- same left-side overlap  length  
- same left-side overlap length   
- same full overlap length  == amplicon lenght      
   
In <span style="color:purple">**purple**</span> it's impossible to find homologous regions.   

For further analysis we need to calculate amplicon and target region length, distance between amplicon and target region start positions, distance between amplicon and target region end positions:

[List]: http://genome.ucsc.edu/cgi-bin/hgSearch?search=pseudogene+mRNA&hgsid=1712866984_aQKAtsLa46m4JJkhGbsoT2fVyJFe&db=hg19

In [14]:
df['ampl_len'] = df[2] - df[1]
df['ref_len'] = df[10] - df[9]
df['ampl_start-ref_start'] = df[1] - df[9]
df['ref_end-ampl_end'] = df[10] - df[2]

df.head(6)

0          1          2               3  4  \
0  chr1  113498778  113499101  AMPL7156752826  .   
1  chr1  113498778  113499101  AMPL7156752826  .   
2  chr1  113498778  113499101  AMPL7156752826  .   
3  chr1  113498778  113499101  AMPL7156752826  .   
4  chr1  113498778  113499101  AMPL7156752826  .   
5  chr1  113498778  113499101  AMPL7156752826  .   

                                        5     6                      7     8  \
0  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  exon   
1  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  exon   
2  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  exon   
3  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  5UTR   
4  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  exon   
5  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  5UTR   

           9         10 11 12 13           14  ampl_len  ref_len  \
0  113499037  113499578  .  +  .  SLC16A1-AS1       323      541   
1  113499037  113500344  .  +  .  SLC16A1-AS1       323     1307   
2  113497898  113498818  .  -  .      SLC16A1       323      920   
3  113497898  113498818  .  -  .      SLC16A1       323      920   
4  113498657  113498818  .  -  .      SLC16A1       323      161   
5  113498657  113498818  .  -  .      SLC16A1       323      161   

   ampl_start-ref_start  ref_end-ampl_end  
0                  -259               477  
1                  -259              1243  
2                   880              -283  
3                   880              -283  
4                   121              -283  
5                   121              -283

If `start_ampl-start_ref` > 0 and `stop_ref-stop_ampl` > 0, it's <span style="color:blue">**blue**</span> position (see picture above)   
If `start_ampl-start_ref` < 0 and `stop_ref-stop_ampl` > 0, it's <span style="color:green">**green**</span> position   
If `start_ampl-start_ref` > 0 and `stop_ref-stop_ampl` < 0, it's <span style="color:yellow">**yellow**</span> position    
If `start_ampl-start_ref` < 0 and `stop_ref-stop_ampl` < 0, it's <span style="color:purple">**purple**</span> position

Now let's calculate overlaps lengths, it will be negative in <span style="color:green">**green**</span> positions, in case of to amplicon with same overlap length in different regions

In [15]:
def overlap_length(df):
    if (df['ampl_start-ref_start'] >= 0) and (df['ref_end-ampl_end'] >= 0):
        return df['ampl_len']
    elif (df['ampl_start-ref_start'] < 0) and (df['ref_end-ampl_end'] > 0):
        return -(df['ampl_len'] + df['ampl_start-ref_start'])
    elif (df['ampl_start-ref_start'] > 0) and (df['ref_end-ampl_end'] < 0):
        return df['ampl_len'] + df['ref_end-ampl_end']
    elif (df['ampl_start-ref_start'] <= 0) and (df['ref_end-ampl_end'] <= 0):
        return np.nan


In [16]:
df['overlap_len'] = df.apply(lambda x: overlap_length(x), axis=1)
df = df.dropna().reset_index().drop('index', axis=1)

We can remove rows with the same amplicon ID, gene, part of gene and overlap length.    
Also we can drop rows where `overlap_len` is unique, because as we notice before, 100% homologu should have same length:

In [17]:
df = df.drop_duplicates([3, 8, 14, 'overlap_len'])

def drop_unique_length(df):
    tmp_list = []
    for x in df['overlap_len'].unique():
        n = list(df['overlap_len']).count(x)
        if n > 1:
            tmp_list.append(x)
    return df.loc[(df['overlap_len'].isin(tmp_list))].reset_index().drop('index', axis=1)


In [18]:
df = drop_unique_length(df)
df.head(5)

0          1          2               3  4  \
0   chr1  113498778  113499101  AMPL7156752826  .   
1   chr1  113498778  113499101  AMPL7156752826  .   
2  chr11    2181834    2182073  AMPL7154784953  .   
3  chr11    2181834    2182073  AMPL7154784953  .   
4  chr11    2181834    2182073  AMPL7154784953  .   

                                                                      5  \
0                                Pool=2;SUBMITTED_REGION=AMPL7156752826   
1                                Pool=2;SUBMITTED_REGION=AMPL7156752826   
2  Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   
3  Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   
4  Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   

       6                      7     8          9         10 11 12 13  \
0   chr1  ncbiRefSeq.2021-05-17  exon  113497898  113498818  .  -  .   
1   chr1  ncbiRefSeq.2021-05-17  5UTR  113497898  113498818  .  -  .   
2  chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182439  .  -  .   
3  chr11  ncbiRefSeq.2021-05-17   CDS    2182015    2182201  .  -  0   
4  chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   

         14  ampl_len  ref_len  ampl_start-ref_start  ref_end-ampl_end  \
0   SLC16A1       323      920                   880              -283   
1   SLC16A1       323      920                   880              -283   
2       INS       239      424                  -181               366   
3       INS       239      186                  -181               128   
4  INS-IGF2       239      203                  -181               145   

   overlap_len  
0         40.0  
1         40.0  
2        -58.0  
3        -58.0  
4        -58.0

Notice, that if we filter only exons, the set of "interesting" genes will be the same, so let's do it:

In [19]:
print(df.loc[df[8] == 'exon'][14].unique())
print(df[14].unique())

df = df.loc[df[8] == 'exon'].reset_index().drop('index', axis=1)

['SLC16A1' 'INS' 'INS-IGF2' 'C12orf43' 'HNF1A']
['SLC16A1' 'INS' 'INS-IGF2' 'C12orf43' 'HNF1A']


And drop rows where `overlap_len` is unique again:

In [20]:
df = drop_unique_length(df)
df

0          1          2               3  4  \
0   chr11    2181834    2182073  AMPL7154784953  .   
1   chr11    2181834    2182073  AMPL7154784953  .   
2   chr11    2182062    2182385  AMPL7154784954  .   
3   chr11    2182062    2182385  AMPL7154784954  .   
4   chr11    2182130    2182458  AMPL7154784955  .   
5   chr11    2182130    2182458  AMPL7154784955  .   
6   chr11    2182429    2182753  AMPL7154897326  .   
7   chr11    2182429    2182753  AMPL7154897326  .   
8   chr12  121438967  121439264  AMPL7154784910  .   
9   chr12  121438967  121439264  AMPL7154784910  .   
10  chr12  121439250  121439535  AMPL7154784911  .   
11  chr12  121439250  121439535  AMPL7154784911  .   
12  chr12  121439524  121439785  AMPL7154784912  .   
13  chr12  121439524  121439785  AMPL7154784912  .   
14  chr12  121439774  121440004  AMPL7154783021  .   
15  chr12  121439774  121440004  AMPL7154783021  .   
16  chr12  121439987  121440308  AMPL7159709172  .   
17  chr12  121439987  121440308  AMPL7159709172  .   

                                                                       5  \
0   Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   
1   Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   
2   Pool=2;SUBMITTED_REGION=AMPL7154784954,AMPL7154784953,AMPL7154784955   
3   Pool=2;SUBMITTED_REGION=AMPL7154784954,AMPL7154784953,AMPL7154784955   
4   Pool=1;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326,AMPL7154784954   
5   Pool=1;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326,AMPL7154784954   
6                  Pool=2;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326   
7                  Pool=2;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326   
8   Pool=1;SUBMITTED_REGION=AMPL7154784910,AMPL7154511317,AMPL7154784911   
9   Pool=1;SUBMITTED_REGION=AMPL7154784910,AMPL7154511317,AMPL7154784911   
10  Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154784910,AMPL7154784911   
11  Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154784910,AMPL7154784911   
12  Pool=1;SUBMITTED_REGION=AMPL7154784911,AMPL7154784912,AMPL7154783021   
13  Pool=1;SUBMITTED_REGION=AMPL7154784911,AMPL7154784912,AMPL7154783021   
14  Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154783021,AMPL7159709172   
15  Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154783021,AMPL7159709172   
16  Pool=1;SUBMITTED_REGION=AMPL7154783021,AMPL7159709205,AMPL7159709172   
17  Pool=1;SUBMITTED_REGION=AMPL7154783021,AMPL7159709205,AMPL7159709172   

        6                      7     8          9         10 11 12 13  \
0   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182439  .  -  .   
1   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   
2   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   
3   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   
4   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   
5   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   
6   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182439  .  -  .   
7   chr11  ncbiRefSeq.2021-05-17  exon    2182398    2182439  .  -  .   
8   chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   
9   chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   
10  chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   
11  chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   
12  chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   
13  chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   
14  chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   
15  chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   
16  chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   
17  chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   

          14  ampl_len  ref_len  ampl_start-ref_start  ref_end-ampl_end  \
0        INS       239      424                  -181               366  

INS-IGF2 is a read-through gene, which overlaps with gene INS at the 5' region, so it's not off-target region.   
Apart from this, there are several amplicons, mapped on two different genes - *HNF1A* and *C12orf43*. Based on the descriptions of these genes (table in task1, file `RESULTS/genes_summary.csv`) I can assume, that *HNF1A* is a homologous off-target region.    
I could be important to notice, that these genes have similar coordinates in different DNA strand. 
   
**Target region coordinates:** 121438868, 121440315    
**Homologous off-target coordinates:** 121438289, 121442292    
   

Save results:

In [21]:
df = df.loc[df[14].isin(['HNF1A', 'C12orf43'])].reset_index().drop('index', axis=1)
df.to_csv('RESULTS/task2_homologous_region.csv', sep='\t', index=None)

*******************************

# 3. Two other ideas how to find homologous off-target region   
3.1 [**Segmental Duplication Database**](https://eeelegacy.gs.washington.edu/humanparalogy/build37/build37.htm)[[3]](https://pubmed.ncbi.nlm.nih.gov/11381028/)     

Download these database and try `bedtools intersect` again


In [22]:
sddb = pd.read_excel("build37.xlsx")

In [23]:
sddb

chrom  chromStart  chromEnd           name  score strand otherChrom  \
0      chr1       85326     87112    chr1:398212      0      _       chr1   
1      chr1      398212    400000     chr1:85326      0      _       chr1   
2      chr1       88000    121417    chr1:235525      0      +       chr1   
3      chr1      235525    267707     chr1:88000      0      +       chr1   
4      chr1       91256     92392    chr1:521369      0      +       chr1   
...     ...         ...       ...            ...    ...    ...        ...   
51594  chrY    58880431  58891092  chrY:58869728      0      +       chrY   
51595  chrY    58881608  58885137  chrY:58885137      0      +       chrY   
51596  chrY    58885137  58888721  chrY:58881608      0      +       chrY   
51597  chrY    59012496  59013612  chrY:59024857      0      _       chrY   
51598  chrY    59024857  59025974  chrY:59012496      0      _       chrY   

       otherStart  otherEnd  otherSize  ...  indelS  alignB  matchB  \
0          398212    400000       1788  ...       2    1786    1757   
1           85326     87112       1786  ...       2    1786    1757   
2          235525    267707      32182  ...    1299   32150   31941   
3           88000    121417      33417  ...    1299   32150   31941   
4          521369    522487       1118  ...      20    1117    1092   
...           ...       ...        ...  ...     ...     ...     ...   
51594    58869728  58880430      10702  ...      99   10632   10566   
51595    58885137  58888721       3584  ...      55    3529    3509   
51596    58881608  58885137       3529  ...      55    3529    3509   
51597    59024857  59025974       1117  ...       3    1115    1041   
51598    59012496  59013612       1116  ...       3    1115    1041   

       mismatchB  transitionsB  transversionsB fracMatch  fracMatchIndel  \
0             29            15              14  0.983763        0.982662   
1             29            15              14  0.983763        0.982662   
2            209           133              76  0.993499        0.992727   
3            209           133              76  0.993499        0.992727   
4             25            18               7  0.977619        0.974130   
...          ...           ...             ...       ...             ...   
51594         66            25              41  0.993792        0.992485   
51595         20             9              11  0.994333        0.993488   
51596         20             9              11  0.994333        0.993488   
51597         74            42              32  0.933632        0.931961   
51598         74            42              32  0.933632        0.931961   

            jcK       k2K  
0      0.016416  0.016423  
1      0.016416  0.016423  
2      0.006529  0.006532  
3      0.006529  0.006532  
4      0.022722  0.022781  
...         ...       ...  
51594  0.006234  0.006234  
51595  0.005689  0.005689  
51596  0.005689  0.005689  
51597  0.069490  0.069706  
51598  0.069490  0.069706  

[51599 rows x 29 columns]

Extract 4-column BED-file:

In [24]:
sddb.rename(columns={'chrom':'#chrom'})
sddb_short = sddb.iloc[:, [0, 1, 2, 3]]
sddb_short.to_csv('sddb.bed', index=None, sep='\t')

Run `bedtools intersect` with SDDB BED-file and output BED-file from tak1 (to have genes names):   
```bash
bedtools intersect -a ./RESULTS/task1_output.bed -b sddb.bed -wa -wb > sddb_intersect.txt
```

In [25]:
pd.read_csv('sddb_intersect.txt', sep='\t', header=None)

0          1          2               3  4   \
0    chr9  135937315  135937599  AMPL7154848171  .   
1    chr9  135939554  135939855  AMPL7154785491  .   
2    chr9  135939844  135940096  AMPL7162958722  .   
3    chr9  135939844  135940096  AMPL7162958722  .   
4    chr9  135940061  135940234  AMPL7154783210  .   
5    chr9  135940223  135940552  AMPL7154785486  .   
6    chr9  135940524  135940722  AMPL7154848184  .   
7    chr9  135941786  135942083  AMPL7154848187  .   
8    chr9  135942100  135942411  AMPL7154848190  .   
9    chr9  135942385  135942705  AMPL7154848193  .   
10   chr9  135943934  135944214  AMPL7154785490  .   
11   chr9  135944183  135944500  AMPL7166072477  .   
12   chr9  135944183  135944500  AMPL7166072477  .   
13   chr9  135944485  135944807  AMPL7166072478  .   
14   chr9  135945694  135946019  AMPL7154785492  .   
15   chr9  135946008  135946189  AMPL7162954932  .   
16   chr9  135946180  135946510  AMPL7154785581  .   
17   chr9  135947121  135947302  AMPL7163451433  .   
18  chr10   88811251   88811541  AMPL7166063029  .   
19  chr10   88811251   88811541  AMPL7166063029  .   
20  chr10   88811251   88811541  AMPL7166063029  .   
21  chr10   88811251   88811541  AMPL7166063029  .   
22  chr10   88811251   88811541  AMPL7166063029  .   
23  chr10   88811251   88811541  AMPL7166063029  .   
24  chr10   88811251   88811541  AMPL7166063029  .   
25  chr10   88811251   88811541  AMPL7166063029  .   

                                                                      5   \
0                                 Pool=1;SUBMITTED_REGION=AMPL7154848171   
1                  Pool=2;SUBMITTED_REGION=AMPL7154785491,AMPL7162958722   
2   Pool=1;SUBMITTED_REGION=AMPL7154783210,AMPL7154785491,AMPL7162958722   
3   Pool=1;SUBMITTED_REGION=AMPL7154783210,AMPL7154785491,AMPL7162958722   
4   Pool=2;SUBMITTED_REGION=AMPL7154785486,AMPL7154783210,AMPL7162958722   
5   Pool=1;SUBMITTED_REGION=AMPL7154785486,AMPL7154783210,AMPL7154848184   
6                  Pool=2;SUBMITTED_REGION=AMPL7154848184,AMPL7154785486   
7                                 Pool=1;SUBMITTED_REGION=AMPL7154848187   
8                  Pool=2;SUBMITTED_REGION=AMPL7154848193,AMPL7154848190   
9                  Pool=1;SUBMITTED_REGION=AMPL7154848190,AMPL7154848193   
10                 Pool=2;SUBMITTED_REGION=AMPL7166072477,AMPL7154785490   
11  Pool=1;SUBMITTED_REGION=AMPL7166072477,AMPL7154785490,AMPL7166072478   
12  Pool=1;SUBMITTED_REGION=AMPL7166072477,AMPL7154785490,AMPL7166072478   
13                 Pool=2;SUBMITTED_REGION=AMPL7166072477,AMPL7166072478   
14                 Pool=1;SUBMITTED_REGION=AMPL7154785492,AMPL7162954932   
15  Pool=2;SUBMITTED_REGION=AMPL7154785581,AMPL7162954932,AMPL7154785492   
16                 Pool=1;SUBMITTED_REGION=AMPL7154785581,AMPL7162954932   
17                                Pool=2;SUBMITTED_REGION=AMPL7163451433   
18                                Pool=1;SUBMITTED_REGION=AMPL7166063029   
19                                Pool=1;SUBMITTED_REGION=AMPL7166063029   
20                                Pool=1;SUBMITTED_REGION=AMPL7166063029   
21                                Pool=1;SUBMITTED_REGION=AMPL7166063029   
22                                Pool=1;SUBMITTED_REGION=AMPL7166063029   
23                                Pool=1;SUBMITTED_REGION=AMPL7166063029   
24                                Pool=1;SUBMITTED_REGION=AMPL7166063029   
25                                Pool=1;SUBMITTED_REGION=AMPL7166063029   

       6   7      8          9          10              11  
0     CEL   1   chr9  135935536  135947286  chr9:135956012  
1     CEL   2   chr9  135935536  135947286  chr9:135956012  
2     CEL   2   chr9  135935536  135947286  chr9:135956012  
3     CEL   3   chr9  135935536  135947286  chr9:135956012  
4     CEL   3   chr9  135935536  135947286  chr9:135956012  
5     CEL   4   chr9  135935536  135947286  chr9:135956012  
6     CEL   4   chr9  135935536  135947286  chr9:135956012  
7     C

We see, that regions in chr10 have duplicates in chrX, for example:

In [26]:
sddb.loc[(sddb['chrom'] == 'chr10') & (sddb['chromStart'] == 88810248)]

chrom  chromStart  chromEnd            name  score strand otherChrom  \
8705  chr10    88810248  88811628  chrX:120183094      0      _       chrX   

      otherStart   otherEnd  otherSize  ...  indelS  alignB  matchB  \
8705   120183094  120184459       1365  ...      31    1357    1308   

      mismatchB  transitionsB  transversionsB fracMatch  fracMatchIndel  \
8705         49            36              13  0.963891         0.95684   

           jcK       k2K  
8705  0.037007  0.037179  

[1 rows x 29 columns]

But there are indels. Perhaps, with a more detailed approach, this database could be useful in searching homologous off-target regions.   

3.2 **Special tools for searching homologous regions**   
There are some tools and pipelines used for searching homologous region, such as [i-ADHoRe](https://academic.oup.com/nar/article/40/2/e11/2409742), [FASTA](https://www.ebi.ac.uk/Tools/sss/fasta/nucleotide.html) or [this algorithm](https://www.nature.com/articles/gim201658)   
   
All of them work with fasta files, we can get it easily from BED-file.   
First of all, we need to get fasta file from input BED-file.  
    
Download reference genome hg19:   
```bash
wget ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/*
for file in ./*.gz; do gunzip $file; done
cat *.fa > hg19.genome.fa
```
   
Getting fasta from BED:   
```bash
bedtools getfasta -fi hg19.genome.fa -bed IAD143293_241_Designed.bed > input.fasta
```
   
I am not testing these tools, but I assume, if we understand their work well, we can adapt them to our needs


*******************